### PR TITLE
Trigger SESSION_EXPIRE event when a timed out session is found

### PR DESCRIPTION
**Pull request by sachiniWettasinghe**
Resolves https://github.com/wso2/product-is/issues/9312

**Description**
With this PR a SESSION_EXPIRE event is fired for a session expiry due to session idle time out or timed out remember-me session. An expired session can be found if a user tries to SSO, log in again or tries to logout from an application after an idle time out.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 1]
```diff
@@ -951,6 +956,79 @@ public static SessionContext getSessionContextFromCache(String key) {
...
+    /**
+     * Retrieve session context from the session cache.
+     *
+     * @param request           HttpServletRequest.
+     * @param context           Authentication context.
+     * @param sessionContextKey Session context key.
+     * @return Session context key.
+     * @throws AuthenticationFailedException Error in triggering session expire event.
+     */
+    public static SessionContext getSessionContextFromCache(HttpServletRequest request, AuthenticationContext context
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:968]: Is it correct to throw the AuthenticationFailedException exception here?
Shouldn't it be some other FrameworkException?

Comment (sachiniWettasinghe) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:967]: Fixed with 6e0671634815499c949ce71f6f8db045d7ad3366

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 2]
```diff
@@ -951,6 +955,79 @@ public static SessionContext getSessionContextFromCache(String key) {
...
+                Map<String, Object> paramMap = new HashMap<>();
+                paramMap.put(FrameworkConstants.AnalyticsAttributes.USER, authenticatedUser);
+                paramMap.put(FrameworkConstants.AnalyticsAttributes.SESSION_ID, context.getSessionIdentifier());
+                Map<String, Object> unmodifiableParamMap = Collections.unmodifiableMap(paramMap);
+                eventProperties.put(IdentityEventConstants.EventProperty.PARAMS, unmodifiableParamMap);
+
+                Event event = new Event(IdentityEventConstants.EventName.SESSION_EXPIRE.name(), eventProperties);
+                eventService.handleEvent(event);
+            } catch (IdentityEventException e) {
+                throw new FrameworkException("Error in triggering the session expire event.", e);
```
Comment (mefarazath) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:1026]: Let's have some context information. Like the sessionId, user etc.

**Review by mefarazath — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 2 above)
Comment (sachiniWettasinghe) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java]: Fixed with cf0c1c45baba0a400fc721abd4eb7e9833acbc4d

**Pull request merged**